packages/astro/src/cli/check/index.ts

```diff
@@ -1,101 +1,360 @@
 
-	diagnostics.forEach((diag) => {
-		diag.diagnostics.forEach((d) => {
-			console.log(printDiagnostic(diag.fileUri, diag.text, d));
+	syncCli: (settings: AstroSettings, options: SyncOptions) => Promise<ProcessExit>;
 
-			switch (d.severity) {
-				case DiagnosticSeverity.Error: {
-					result.errors++;
-					break;
-				}
-				case DiagnosticSeverity.Warning: {
-					result.warnings++;
-					break;
-				}
-				case DiagnosticSeverity.Hint: {
-					result.hints++;
-					break;
-				}
+	settings: Readonly<AstroSettings>;
+
+	logging: Readonly<LogOptions>;
+
+	fileSystem: typeof fsMod;
+};
+
+/**
+ * Responsible to check files - classic or watch mode - and report diagnostics.
+ *
+ * When in watch mode, the class does a whole check pass, and then starts watching files.
+ * When a change occurs to an `.astro` file, the checker builds content collections again and lint all the `.astro` files.
+ */
+export class AstroChecker {
+	readonly #diagnosticsChecker: AstroCheck;
+	readonly #shouldWatch: boolean;
+	readonly #syncCli: (settings: AstroSettings, opts: SyncOptions) => Promise<ProcessExit>;
+
+	readonly #settings: AstroSettings;
+
+	readonly #logging: LogOptions;
+	readonly #fs: typeof fsMod;
+	#watcher?: FSWatcher;
+
+	#filesCount: number;
+	#updateDiagnostics: NodeJS.Timeout | undefined;
+	constructor({
+		diagnosticChecker,
+		isWatchMode,
+		syncCli,
+		settings,
+		fileSystem,
+		logging,
+	}: CheckerConstructor) {
+		this.#diagnosticsChecker = diagnosticChecker;
+		this.#shouldWatch = isWatchMode;
+		this.#syncCli = syncCli;
+		this.#logging = logging;
+		this.#settings = settings;
+		this.#fs = fileSystem;
+		this.#filesCount = 0;
+	}
+
+	/**
+	 * Check all `.astro` files once and then finishes the operation.
+	 * @returns {Promise<CheckResult>}
+	 */
+	public async check(): Promise<CheckResult> {
+		return await this.#checkAllFiles(true);
+	}
+
+	/**
+	 * Check all `.astro` files and then start watching for changes.
+	 * @returns {Promise<CheckResult>}
+	 */
+	public async watch(): Promise<CheckResult> {
+		await this.#checkAllFiles(true);
+		await this.#watch();
+		return CheckResult.Listen;
+	}
+
+	/**
+	 * Stops the watch. It terminates the inner server.
+	 */
+	public async stop() {
+		await this.#watcher?.close();
+	}
+
+	/**
+	 * Whether the checker should run in watch mode
+	 * @returns {boolean}
+	 */
+	public get isWatchMode(): boolean {
+		return this.#shouldWatch;
+	}
+
+	async #openDocuments() {
+		this.#filesCount = await openAllDocuments(
+			this.#settings.config.root,
+			[],
+			this.#diagnosticsChecker
+		);
+	}
+
+	/**
+	 * Lint all `.astro` files, and report the result in console. Operations executed, in order:
+	 * 1. Compile content collections.
+	 * 2. Optionally, traverse the file system for `.astro` files and saves their paths.
+	 * 3. Get diagnostics for said files and print the result in console.
+	 *
+	 * @param openDocuments Whether the operation should open all `.astro` files
+	 * @private
```

Comment:
I'm not sure if this `@private` is needed as the function is already prefixed with `#`. The other private functions doesn't have this annotation too.